typed-racket-lib/typed-racket/typecheck/tc-expr-unit.rkt

```diff
@@ -398,14 +398,28 @@
        [_ (-box (generalize (find-stx-type x)))])]
     [(? hash? h)
      (match (and expected (resolve (intersect expected -HashTop)))
-       [(Hashtable: kt vt)
+       [(and ht (or (Mutable-Hashtable: kt vt) (Immutable-Hashtable: kt vt) (Weak-Hashtable: kt vt)))
         (define kts (hash-map h (lambda (x y) (find-stx-type x kt))))
         (define vts (hash-map h (lambda (x y) (find-stx-type y vt))))
-        (make-Hashtable
+        (define make-H
+          (cond
+           [(or (Immutable-Hashtable? ht)
+                (immutable? h))
+            make-Immutable-Hashtable]
+           [(or (Weak-Hashtable? ht)
+                (hash-weak? h))
+            make-Weak-Hashtable]
+           [else ;(Mutable-Hashtable? ht)
+            make-Mutable-Hashtable]))
+        (make-H
          (check-below (apply Un kts) kt)
          (check-below (apply Un vts) vt))]
-       [_ (make-Hashtable (generalize (apply Un (map find-stx-type (hash-keys h))))
-                          (generalize (apply Un (map find-stx-type (hash-values h)))))])]
+       [_
+        (let ([kt (generalize (apply Un (map find-stx-type (hash-keys h))))]
+              [vt (generalize (apply Un (map find-stx-type (hash-values h))))])
+          (Un (make-Mutable-Hashtable kt vt)
```

Comment:
Yes, there are lots of potential places, but you could just put it in `tc-literal`.